test/distributed/org/apache/cassandra/fuzz/snapshots/SnapshotsTest.java

```diff
@@ -0,0 +1,880 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.fuzz.snapshots;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+
+import com.google.common.collect.MapDifference;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import org.junit.Assert;
+import org.junit.Test;
+
+import accord.utils.Property.Command;
+import accord.utils.RandomSource;
+import org.apache.cassandra.distributed.Cluster;
+import org.apache.cassandra.distributed.api.Feature;
+import org.apache.cassandra.distributed.api.IInvokableInstance;
+import org.apache.cassandra.distributed.api.IIsolatedExecutor;
+import org.apache.cassandra.distributed.api.IIsolatedExecutor.SerializableTriConsumer;
+import org.apache.cassandra.distributed.api.NodeToolResult;
+import org.apache.cassandra.fuzz.snapshots.SnapshotsTest.State.TestSnapshot;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.schema.KeyspaceMetadata;
+import org.apache.cassandra.schema.Keyspaces;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.service.snapshot.SnapshotManager;
+import org.apache.cassandra.service.snapshot.TableSnapshot;
+import org.apache.cassandra.utils.Generators;
+import org.apache.cassandra.utils.Pair;
+import org.quicktheories.core.RandomnessSource;
+import org.quicktheories.impl.JavaRandom;
+
+import static accord.utils.Property.commands;
+import static accord.utils.Property.stateful;
+import static com.google.common.collect.Sets.difference;
+import static java.lang.String.format;
+import static org.junit.Assert.assertEquals;
+
+public class SnapshotsTest
+{
+    @Test
+    public void fuzzySnapshotsTest()
+    {
+        stateful()
+        .withExamples(1)
+        .withSteps(1000)
+        .check(commands(() -> State::new, state -> new SnapshotsSut(state).start())
+               .add(10, new CreateKeyspaceCommand())
```

Comment:
there is `addIf` that would let you block creating the keyspace if there are too many

Reply:
```suggestion
               .add((rs, state) -> new CreateKeyspaceCommand(state.keyspaceName.next(rs), state.nodeGen.nextInt(rs)))
```

its best practice to define your command state *before* the command is created, this makes the name easier, and also makes the `apply` logic cleaner.  

When you merge everything into the run method then you have to start to save state into this object and you *hope* that the framework calls you in the same order (its free to call `run` before `apply`).

Reply:
The logic I have is too complicated to put the state outside / before command creation. I'll keep it like it is for now.